File: surf-api-core/surf-api-core-api/src/main/java/dev/slne/surf/surfapi/core/api/util/ParticleFactory.java
package dev.slne.surf.surfapi.core.api.util;

import com.github.retrooper.packetevents.protocol.item.ItemStack;
import com.github.retrooper.packetevents.protocol.particle.Particle;
import com.github.retrooper.packetevents.protocol.particle.data.*;
import com.github.retrooper.packetevents.protocol.particle.type.ParticleType;
import com.github.retrooper.packetevents.protocol.world.states.WrappedBlockState;
import com.github.retrooper.packetevents.util.Vector3i;
import net.kyori.adventure.text.format.TextColor;
import org.jetbrains.annotations.ApiStatus;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import static com.google.common.base.Preconditions.*;

@ApiStatus.NonExtendable
public interface ParticleFactory {

    @Contract("_ -> new")
    static @NotNull Particle of(@NotNull ParticleType type) {
        checkNotNull(type, "Particle type may not be null");

        return new Particle(type);
    }

    @Contract(value = "_, _ -> new", pure = true)
    static @NotNull Particle of(@NotNull ParticleType type, @NotNull ParticleData data) {
        checkNotNull(type, "Particle type may not be null");
        checkNotNull(data, "Particle data may not be null");

        return new Particle(type, data);
    }

    @Contract("_, _ -> new")
    static @NotNull Particle of(@NotNull ParticleType type, @NotNull WrappedBlockState blockState) {
        checkNotNull(blockState, "Block state may not be null");

        return of(type, new ParticleBlockStateData(blockState));
    }

    @Contract("_, _, _, _ -> new")
    static @NotNull Particle of(@NotNull ParticleType type, float scale, @NotNull TextColor start, @NotNull TextColor end) {
        checkNotNull(start, "Start color may not be null");
        checkNotNull(end, "End color may not be null");

        return of(type, new ParticleDustColorTransitionData(scale, start.red(), start.green(), start.blue(), end.red(), end.green(), end.blue()));
    }

    @Contract("_, _ -> new")
    static @NotNull Particle of(@NotNull ParticleType type, @NotNull ItemStack itemStack) {
        checkNotNull(itemStack, "Item stack may not be null");

        return of(type, new ParticleItemStackData(itemStack));
    }

    @Contract("_, _ -> new")
    static @NotNull Particle of(@NotNull ParticleType type, float roll) {
        return of(type, new ParticleSculkChargeData(roll));
    }

    @Contract("_, _ -> new")
    static @NotNull Particle of(@NotNull ParticleType type, int delay) {
        return of(type, new ParticleShriekData(delay));
    }

    @Contract("_, _, _, _ -> new")
    static @NotNull Particle of(@NotNull ParticleType type, @NotNull Vector3i startingPosition, @Nullable Vector3i blockPosition, int ticks) {
        checkNotNull(startingPosition, "Starting position may not be null");

        return of(type, new ParticleVibrationData(startingPosition, blockPosition, ticks));
    }

    @Contract("_, _, _, _ -> new")
    static @NotNull Particle of(@NotNull ParticleType type, @NotNull Vector3i startingPosition, int entityId, int ticks) {
        checkNotNull(startingPosition, "Starting position may not be null");

        return of(type, new ParticleVibrationData(startingPosition, entityId, ticks));
    }
}

File: surf-api-core/surf-api-core-api/src/main/kotlin/dev/slne/surf/surfapi/core/api/util/ParticleFactory.kt
package dev.slne.surf.surfapi.core.api.util

import com.github.retrooper.packetevents.protocol.item.ItemStack
import com.github.retrooper.packetevents.protocol.particle.Particle
import com.github.retrooper.packetevents.protocol.particle.data.*
import com.github.retrooper.packetevents.protocol.particle.type.ParticleType
import com.github.retrooper.packetevents.protocol.world.states.WrappedBlockState
import com.github.retrooper.packetevents.util.Vector3i
import net.kyori.adventure.text.format.TextColor
import org.jetbrains.annotations.ApiStatus

@ApiStatus.NonExtendable
@Deprecated("Not longer maintained.")
interface ParticleFactory {
    companion object {
        @JvmStatic
        @Deprecated(
            "Not longer maintained.",
            ReplaceWith("Particle(type)", "com.github.retrooper.packetevents.protocol.particle.Particle")
        )
        fun of(type: ParticleType<*>): Particle<*> {
            return Particle(type)
        }

        @Deprecated(
            "Not longer maintained.",
            ReplaceWith("Particle(type, data)", "com.github.retrooper.packetevents.protocol.particle.Particle")
        )
        fun <D : ParticleData> of(type: ParticleType<D>, data: D): Particle<D> {
            return Particle(type, data)
        }

        @Deprecated(
            "Not longer maintained.",
            ReplaceWith(
                "Particle(type, ParticleBlockStateData(blockState))",
                "com.github.retrooper.packetevents.protocol.particle.Particle",
                "com.github.retrooper.packetevents.protocol.particle.data.ParticleBlockStateData.ParticleBlockStateData"
            )
        )
        fun of(
            type: ParticleType<ParticleBlockStateData>,
            blockState: WrappedBlockState
        ): Particle<ParticleBlockStateData> {
            return Particle(type, ParticleBlockStateData(blockState))
        }

        @Deprecated(
            "Not longer maintained.",
            ReplaceWith(
                "Particle(type, ParticleDustColorTransitionData(scale, start.red().toFloat(), start.green().toFloat(), start.blue().toFloat(), end.red().toFloat(), end.green().toFloat(), end.blue().toFloat()))",
                "com.github.retrooper.packetevents.protocol.particle.Particle",
                "com.github.retrooper.packetevents.protocol.particle.data.ParticleDustColorTransitionData",
            )
        )
        fun of(
            type: ParticleType<ParticleDustColorTransitionData>, scale: Float, start: TextColor,
            end: TextColor
        ): Particle<ParticleDustColorTransitionData> {
            return Particle(
                type,
                ParticleDustColorTransitionData(
                    scale, start.red().toFloat(), start.green().toFloat(), start.blue().toFloat(),
                    end.red().toFloat(), end.green().toFloat(), end.blue().toFloat()
                )
            )
        }

        @Deprecated(
            "Not longer maintained.",
            ReplaceWith(
                "Particle(type, ParticleItemStackData(itemStack))",
                "com.github.retrooper.packetevents.protocol.particle.Particle",
                "com.github.retrooper.packetevents.protocol.particle.data.ParticleItemStackData"
            )
        )
        fun of(
            type: ParticleType<ParticleItemStackData>,
            itemStack: ItemStack
        ): Particle<ParticleItemStackData> {
            return Particle(type, ParticleItemStackData(itemStack))
        }

        @Deprecated(
            "Not longer maintained.",
            ReplaceWith(
                "Particle(type, ParticleSculkChargeData(roll))",
                "com.github.retrooper.packetevents.protocol.particle.Particle",
                "com.github.retrooper.packetevents.protocol.particle.data.ParticleSculkChargeData"
            )
        )
        fun of(
            type: ParticleType<ParticleSculkChargeData>,
            roll: Float
        ): Particle<ParticleSculkChargeData> {
            return Particle(type, ParticleSculkChargeData(roll))
        }

        @Deprecated(
            "Not longer maintained.",
            ReplaceWith(
                "Particle(type, ParticleShriekData(delay))",
                "com.github.retrooper.packetevents.protocol.particle.Particle",
                "com.github.retrooper.packetevents.protocol.particle.data.ParticleShriekData"
            )
        )
        fun of(type: ParticleType<ParticleShriekData>, delay: Int): Particle<ParticleShriekData> {
            return Particle(type, ParticleShriekData(delay))
        }

        @Deprecated(
            "Not longer maintained.",
            ReplaceWith(
                "Particle(type, ParticleVibrationData(startingPosition, blockPosition, ticks))",
                "com.github.retrooper.packetevents.protocol.particle.Particle",
                "com.github.retrooper.packetevents.protocol.particle.data.ParticleVibrationData"
            )
        )
        fun of(
            type: ParticleType<ParticleVibrationData>, startingPosition: Vector3i,
            blockPosition: Vector3i, ticks: Int
        ): Particle<ParticleVibrationData> {
            return Particle(type, ParticleVibrationData(startingPosition, blockPosition, ticks))
        }

        @Deprecated(
            "Not longer maintained.",
            ReplaceWith(
                "Particle(type, ParticleVibrationData(startingPosition, entityId, ticks))",
                "com.github.retrooper.packetevents.protocol.particle.Particle",
                "com.github.retrooper.packetevents.protocol.particle.data.ParticleVibrationData"
            )
        )
        fun of(
            type: ParticleType<ParticleVibrationData>, startingPosition: Vector3i,
            entityId: Int, ticks: Int
        ): Particle<ParticleVibrationData> {
            return Particle(type, ParticleVibrationData(startingPosition, entityId, ticks))
        }
    }
}
